src/Coverage.TestLogger/CoverageLogger.cs

```diff
@@ -0,0 +1,107 @@
+
+    /// <summary>
+    /// Logger for Generating CodeCoverage Analysis
+    /// </summary>
+    [FriendlyName(FriendlyName)]
+    [ExtensionUri(CoverageUri)]
+    internal class CoverageLogger : ITestLogger
+    {
+        #region Constants
+
+        private const string CoverageUri = "datacollector://microsoft/CodeCoverage/2.0";
+
+        private const string FriendlyName = "CoverageLogger";
+
+        private const string SetupInteropx86 = @"x86\Microsoft.VisualStudio.Setup.Configuration.Native.dll";
+
+        private const string SetupInteropx64 = @"x64\Microsoft.VisualStudio.Setup.Configuration.Native.dll";
+
+        private static readonly Uri CodeCoverageDataCollectorUri = new Uri(CoverageUri);
+
+        private ManualResetEvent coverageXmlGenerateEvent;
+
+        private CodeCoverageUtility codeCoverageUtility;
+
+        #endregion
+
+        #region ITestLogger
+
+        /// <inheritdoc/>
+        public void Initialize(TestLoggerEvents events, string testResultsDirPath)
+        {
+            if (events == null)
+            {
+                throw new ArgumentNullException(nameof(events));
+            }
+
+            // Register for the events.
+            events.TestRunComplete += this.TestRunCompleteHandler;
+
+            this.codeCoverageUtility = new CodeCoverageUtility();
+        }
+        #endregion
+
+        #region Event Handlers
+
+        /// <summary>
+        /// Called when a test run is completed.
+        /// </summary>
+        /// <param name="sender">
+        /// The sender.
+        /// </param>
+        /// <param name="e">
+        /// Test run complete events arguments.
+        /// </param>
+        internal void TestRunCompleteHandler(object sender, TestRunCompleteEventArgs e)
+        {
+            if (e.AttachmentSets == null)
+            {
+                return;
+            }
+
+            var coverageAttachments = e.AttachmentSets
+                .Where(dataCollectionAttachment => CodeCoverageDataCollectorUri.Equals(dataCollectionAttachment.Uri)).ToArray();
+
+            if (coverageAttachments.Any())
+            {
+                var codeCoverageFiles = coverageAttachments.Select(coverageAttachment => coverageAttachment.Attachments[0].Uri.LocalPath).ToArray();
+                foreach (var codeCoverageFile in codeCoverageFiles)
+                {
+                    var resultFile = Path.Combine(Path.GetDirectoryName(codeCoverageFile), Path.GetFileNameWithoutExtension(codeCoverageFile) + ".xml");
+                    var arguments = "analyze /output:" + '"' + resultFile + '"' + " " + '"' + codeCoverageFile + '"';
+                    try
+                    {
+                        this.codeCoverageUtility.AnalyzeCoverageFile(arguments, this.GetCodeCoverageExePath());
+                    }
+                    catch (Exception ex)
+                    {
+                        ConsoleOutput.Instance.WriteLine(ex.Message, OutputLevel.Information);
```

Comment:
Is ConsoleOutput coming from TP? How and when did this become public? :godmode:

Reply:
Let's not take this dependency if possible. It shouldn't be a publicly documented api somehow.

Reply:
Used Console.WriteLine for now.